### PR TITLE
Client specific pipelines.

### PR DESCRIPTION
**Pull request by mitchdenny**
This PR lays down the configuration for six seperate PR pipelines which build:

- sdk/servicebus
- sdk/eventhub
- sdk/opertionalinsights
- sdk/hdinsight
- sdk/keyvault
- sdk/all (triggers when files in eng and .azure-pipelines are updated)

Once these are merged I'll setup six pipelines in the public project which point to the ```pr.client.yml``` files in this PR. These pipelines will then become the primary validation mechanism for PRs into the folders in the list above. Other pipelines that are setup to trigger on these folders should be removed once we've confirmed that these pipelines cover everything they need.

Each of the ```pr.client.yml``` files references a job template ```archetype-sdk-client.yml```. This is basically a copy of what is in ```client.yml``` in ```.azure-pipelines```. The idea is that eventually we'll also create a CI build which builds all of the data plane which uses the same definition (that will be a future PR).

This change should reduce average PR build times for teams by about 60% depending on the code-base being built.

### CODE REVIEW

File: sdk/all/pr.client.yml
Comment (weshaggard) [inline sdk/all/pr.client.yml]: Is it necessary to specify the branches?

Comment (mitchdenny) [reply sdk/all/pr.client.yml]: No I'll remove that in a subsequent PR. I was using this for testing purposes as I was debugging the script. I can remove now if you would prefer.

Comment (mitchdenny) [reply sdk/all/pr.client.yml]: I went ahead and removed it because as soon as we merge this PR I'll end up putting in another straight away to remove it - may as well do it now.

File: sdk/eventhub/.azure-pipelines/pr.client.yml
Comment (weshaggard) [inline sdk/eventhub/.azure-pipelines/pr.client.yml]: While these paths are OK for now we will have to think about if we want to scope to just data-plane once we move over the management libraries under the same directory.

File: sdk/eventhub/.azure-pipelines/pr.client.yml
Comment (weshaggard) [inline sdk/eventhub/.azure-pipelines/pr.client.yml]: Do you think it is worth scoping these under a ".azure-pipelines" folder? I'm not necessary against it but I'm curious if there is a reason you had in mind. On potential issue I can see is that we might need to add some exceptions to some of or tools as they are going to be assuming each directory under the service folder is a shipping package.

File: .azure-pipelines/templates/jobs/archetype-sdk-client.yml
Comment (weshaggard) [inline .azure-pipelines/templates/jobs/archetype-sdk-client.yml]: There isn't any need to add extra escaping for the variable references? Or is that only for the template parameters?

File: .azure-pipelines/templates/variables/globals.yml
Comment (weshaggard) [inline .azure-pipelines/templates/variables/globals.yml]: Is there any reason this cannot go into archetype-sdk-client.yml?

File: sdk/keyvault/.azure-pipelines/pr.client.yml
Comment (weshaggard) [inline sdk/keyvault/.azure-pipelines/pr.client.yml]: Did you actually generate these or did you copy/paste them? At any case I would remove the "may be" part.

File: sdk/keyvault/.azure-pipelines/pr.client.yml
Comment (weshaggard) [inline sdk/keyvault/.azure-pipelines/pr.client.yml]: Is this pr section necessary? master is the default correct?

**Comment by weshaggard** [comment]
I left some comments but overall this looks reasonable. What is the naming convention you plan to use for the pipelines? I suspect we should work with @danieljurek with test pipelines and come up with a smaller naming convention given how many we are going to have.

File: sdk/eventhub/.azure-pipelines/pr.client.yml
Comment (mitchdenny) [reply sdk/eventhub/.azure-pipelines/pr.client.yml]: Yep ... I think we'll end up with ```pr.client.yml``` referencing a set of paths, and if the ADX team wants to do something similar they might have a ```pr.mgmt.yml``` with another set, possibly with some overlap if there are common dependencies.

File: .azure-pipelines/templates/jobs/archetype-sdk-client.yml
Comment (mitchdenny) [reply .azure-pipelines/templates/jobs/archetype-sdk-client.yml]: Not sure what you mean? I know that sometimes you need to surround a value with quotes IF it breaks the YAML parsing, but this usage seems OK to me. Are you referring to vmIMage or the DotNetCoreVersion?

File: .azure-pipelines/templates/variables/globals.yml
Comment (mitchdenny) [reply .azure-pipelines/templates/variables/globals.yml]: Generally agree. I'll need to test whether I can import templates into the jobs template and have it impact all jobs within that template.

File: sdk/keyvault/.azure-pipelines/pr.client.yml
Comment (mitchdenny) [reply sdk/keyvault/.azure-pipelines/pr.client.yml]: I hand wrote them but included the warning to dissuade modifying them :P I'll make it more explicity even though they aren't automatically generated ... yet.

File: sdk/eventhub/.azure-pipelines/pr.client.yml
Comment (mitchdenny) [reply sdk/eventhub/.azure-pipelines/pr.client.yml]: I think we can take two approaches, one is to have a structure like the following:

```
.azure-pipelines/
  components/
    servicebus/
      pr.client.yml
  templates/
    jobs/
      archetype-sdk-client.yml
sdk/
  servicebus/
    (actual client code)
```

or something like this (current PR):

```
.azure-pipelines/
  templates/
    jobs/
      archetype-sdk-client.yml
sdk/
  servicebus/
    .azure-pipelines/
      pr.client.yml
    (actual client code)
```

One of the pros for the first approach is that it puts all the pipeline related configuration under one folder, and probably discourages changes by devs. I've gone with the second approach for now but happy to flick it back. I don't have a strong preference at this point.

File: .azure-pipelines/templates/jobs/archetype-sdk-client.yml
Comment (weshaggard) [reply .azure-pipelines/templates/jobs/archetype-sdk-client.yml]: I was meaning like `${{ parameters.name }}` I thought all variables needed to be enclosed like that but perhaps only parameters. I need to spend some more time playing with templates so I can better understand how they are evaluated.

File: sdk/eventhub/.azure-pipelines/pr.client.yml
Comment (weshaggard) [reply sdk/eventhub/.azure-pipelines/pr.client.yml]: I like having all the things related to a given service under the service directory that wasn't my question. I was more wondering why you did it under `sdk/servicebus/.azure-pipelines/pr.client.yml` vs `sdk/servicebus/pr.client.yml`

Comment (mitchdenny) [reply sdk/eventhub/.azure-pipelines/pr.client.yml]: Oh - that was just in case we need to have seperate files in there (just in case).

File: .azure-pipelines/templates/jobs/archetype-sdk-client.yml
Comment (mitchdenny) [reply .azure-pipelines/templates/jobs/archetype-sdk-client.yml]: I think that is only when you want to propogate variables between jobs that didn't exist at the time the YAML file was processed. For example, if I have a task that produces a variable (calls ##vso[setvariable..]) then that variable is valid across steps within the same job - but to jump across jobs I need to specify that as an output variable for the job - I think.

Comment (mitchdenny) [reply .azure-pipelines/templates/jobs/archetype-sdk-client.yml]: In short - I think this usage is OK.

File: sdk/keyvault/.azure-pipelines/pr.client.yml
Comment (mitchdenny) [reply sdk/keyvault/.azure-pipelines/pr.client.yml]: I'm going to test to make sure that master is the default and that its not a wildcard, once I've checked into that I'll make the necessary modifications. The documentation is a little ambiguous.

File: sdk/eventhub/.azure-pipelines/pr.client.yml
Comment (mitchdenny) [reply sdk/eventhub/.azure-pipelines/pr.client.yml]: After some thought I am going to remove the .azure-pipelines folders under each service directory. We can add it back in if we ever have the need but its not necessary right now since in most cases the files themselves should be all that is needed.

File: sdk/keyvault/.azure-pipelines/pr.client.yml
Comment (mitchdenny) [reply sdk/keyvault/.azure-pipelines/pr.client.yml]: So the ```pr:``` statement is redundant in that PR validation is the default behaviour, but it is the default behavior on all branches. This may or may not be desirable. For the time being I am going to remove it as unnecessary but we should consider whether we want to force these pipelines to run when folks want to merge between branches other than master (not a bad thing on the surface but could be annoying in some peoples workflows).

File: .azure-pipelines/templates/variables/globals.yml
Comment (mitchdenny) [reply .azure-pipelines/templates/variables/globals.yml]: OK - what I've done is I've kept a seperate file (I think having a bunch of these settings seperately might be useful), however I now reference that template directly within the ```archetype-sdk-client``` template so the global variables don't leak up in into the ```pr.client.yml``` definitions.

**Comment by mitchdenny** [comment]
This PR is now ready to go/merge pending final review. I've responded to suggestions and I've validated that the revisions work in the playground pipelines that I've setup.

**Review by weshaggard — APPROVED** [review]

File: sdk/all/pr.client.yml
Comment (weshaggard) [inline sdk/all/pr.client.yml]: Just noticed the typo, need to remove "be"

Comment (mitchdenny) [reply sdk/all/pr.client.yml]: Fixed.

**Pull request merged**
